engine/render/renderer/platform/vulkan/Font.h

```diff
@@ -0,0 +1,198 @@
+//      https://opensource.org/licenses/Zlib
+//
+
+#ifndef SIEGE_ENGINE_FONT_H
+#define SIEGE_ENGINE_FONT_H
+
+#include <freetype/freetype.h>
+#include <utils/Hash.h>
+#include <utils/collections/StackArray.h>
+
+#include "Texture2D.h"
+
+namespace Siege::Vulkan
+{
+
+/**
+ * The Font class is responsible for loading in fonts from disk. Fonts are loaded into a texture
+ * atlas with every character represented as a glyph. The Font class controls how these glyphs are
+ * loaded in and stored
+ */
+class Font
+{
+public:
+
+    // Constexpr declarations
+    static constexpr uint32_t INITIAL_FONT_ATLAS_WIDTH {512};
+    static constexpr uint32_t INITIAL_FONT_ATLAS_HEIGHT {512};
+
+    // Structs
+
+    /**
+     * Stores information about the texture atlas held by the font. This is used for rendering
+     */
+    struct Info
+    {
+        Texture2D::Info textureInfo {};
+        VkSampler sampler {nullptr};
+    };
+
+    /**
+     * A glyph is a single character (currently only ASCII characters). Each character takes up a
+     * specific region of space in the texture atlas. The glyph struct stores information such as
+     * the glyph's uv coordinates and dimensions
+     */
+    struct Glyph
+    {
+        float uvxMin {0};
+        float uvyMin {0};
+        float uvxMax {0};
+        float uvyMax {0};
+        float width {0};
+        float height {0};
+        float widthNormalised {0};
+        float heightNormalised {0};
+        float bearingX {0};
+        float bearingY {0};
+        unsigned int advance {0};
+    };
+
+    // 'Structors
+
+    /**
+     * A default constructor for the Font class
+     */
+    Font() = default;
+
+    /**
+     * A base constructor. This constructor loads in a font from a specific location on disk
+     * @param filePath the file path to load the font from
+     */
+    Font(const char* filePath);
+
+    /**
+     * Swaps the contents of two Font objects
+     * @param other the Font object to swap values with
+     */
+    inline Font(Font&& other)
+    {
+        Swap(other);
+    }
+
+    /**
+     * The Font destructor. Frees all memory held by the Font class
+     */
+    ~Font();
+
+    // Operators
+
+    /**
+     * The Font move assignment operator. Creates a new Font and moves it's contents into the
+     * current Font
+     * @param other the Font to move data into
+     * @return a reference to the current Font object
+     */
+    inline Font& operator=(Font&& other)
+    {
+        Swap(other);
+        return *this;
+    }
+
+    // Public methods
+
+    /**
+     * Frees the memory held by the Font
+     */
+    void Free();
+
+    /**
+     * Initialises the dependent libs (FreeType). This function should only be called once when the
+     * engine starts up
+     */
+    static void InitialiseFontLibs();
+
+    /**
+     * Destroys the dependent font libs. This function should only be called when the engine shuts
+     * down
+     */
+    static void DestroyFontLibs();
```

Comment:
Do these all need to be public or can we make them friend classes for something because it seems they're not supposed to be called outside of certain engine lifetime events